openfermioncirq/variational/ansatzes/low_rank.py

```diff
@@ -0,0 +1,303 @@
+        self.adiabatic_evolution_time = cast(float, adiabatic_evolution_time)
+
+        # Perform the low rank decomposition of two-body operator.
+        self.eigenvalues, one_body_squares, self.one_body_correction, _ = (
+            low_rank_two_body_decomposition(
+                hamiltonian.two_body_tensor,
+                final_rank=self.final_rank,
+                spin_basis=spin_basis))
+        self.one_body_coefficients = (
+                hamiltonian.one_body_tensor + self.one_body_correction)
+        self.constant = hamiltonian.constant
+
+        # Get scaled density-density terms and basis transformation matrices.
+        self.scaled_density_density_matrices = []  # type: List[numpy.ndarray]
+        self.basis_change_matrices = []            # type: List[numpy.ndarray]
+        for j in range(len(self.eigenvalues)):
+            density_density_matrix, basis_change_matrix = (
+                prepare_one_body_squared_evolution(one_body_squares[j]))
+            self.scaled_density_density_matrices.append(
+                    numpy.real(self.eigenvalues[j] * density_density_matrix))
+            self.basis_change_matrices.append(basis_change_matrix)
+
+        super().__init__(qubits)
+
+    def params(self) -> Iterable[cirq.Symbol]:
+        """The parameters of the ansatz."""
+
+        for i in range(self.iterations):
+
+            for p in range(len(self.qubits)):
+                # Diagonal one-body coefficients
+                if (self.include_all_z or not numpy.isclose(
+                        self.one_body_coefficients[p, p], 0)):
+                    yield LetterWithSubscripts('U', p, i)
+                # Diagonal two-body coefficients for each singular vector
+                for j in range(len(self.eigenvalues)):
+                    two_body_coefficients = (
+                            self.scaled_density_density_matrices[j])
+                    if (self.include_all_z or not numpy.isclose(
+                            two_body_coefficients[p, p], 0)):
+                        yield LetterWithSubscripts('U', p, j, i)
+
+            for p, q in itertools.combinations(range(len(self.qubits)), 2):
+                # Off-diagonal one-body coefficients
+                if (self.include_all_xxyy or not numpy.isclose(
+                        self.one_body_coefficients[p, q].real, 0)):
+                    yield LetterWithSubscripts('T', p, q, i)
+                # Off-diagonal two-body coefficients for each singular vector
+                for j in range(len(self.eigenvalues)):
+                    two_body_coefficients = (
+                            self.scaled_density_density_matrices[j])
+                    if (self.include_all_cz or not numpy.isclose(
+                            two_body_coefficients[p, q], 0)):
+                        yield LetterWithSubscripts('V', p, q, j, i)
+
+    def param_bounds(self) -> Optional[Sequence[Tuple[float, float]]]:
```

Comment:
where are these bounds coming from?

Reply:
The `U` and `V` parameters correspond to Z rotations so they have a period of 2 half-turns (2pi radians). The `T` parameters correspond to XXYY rotations so they have a period of 4 half-turns. #267 gets rid of the `T` parameters.

Reply:
What would I care about the period? You are evolving under a Hamiltonian. Please just express the evolution like that. You're not exponentiating a gate. You're evolving under a Hamiltonian. That's all we do in quantum simulation.

Reply:
Some optimization algorithms require bounds on the inputs. And indeed, restricting the parameters to within these bounds results in no loss of generality; there exist a set of parameters within these bounds which are optimal.